src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs

```diff
@@ -0,0 +1,634 @@
+        [Fact]
+        public void SwitchExpression_LocalsClash()
+        {
+            string source = @"
+using System;
+class X
+{
+    void M(int? x, object y)
+    {
+        y = /*<bind>*/x switch { int z when x is int z => 5 }/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+ISwitchExpressionOperation (1 arms) (OperationKind.SwitchExpression, Type: System.Int32, IsInvalid) (Syntax: 'x switch {  ... nt z => 5 }')
+  Value: 
+    IParameterReferenceOperation: x (OperationKind.ParameterReference, Type: System.Int32?) (Syntax: 'x')
+  Arms(1):
+      ISwitchExpressionArmOperation (2 locals) (OperationKind.SwitchExpressionArm, Type: null, IsInvalid) (Syntax: 'int z when  ...  int z => 5')
+        Pattern: 
+          IDeclarationPatternOperation (OperationKind.DeclarationPattern, Type: null) (Syntax: 'int z') (InputType: System.Int32?, DeclaredSymbol: System.Int32 z, MatchesNull: False)
+        Guard: 
+          IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Boolean, IsInvalid, IsImplicit) (Syntax: 'x is int z')
+            Conversion: CommonConversion (Exists: False, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+            Operand: 
+              IIsPatternOperation (OperationKind.IsPattern, Type: System.Boolean, IsInvalid) (Syntax: 'x is int z')
+                Value: 
+                  IParameterReferenceOperation: x (OperationKind.ParameterReference, Type: System.Int32?) (Syntax: 'x')
+                Pattern: 
+                  IDeclarationPatternOperation (OperationKind.DeclarationPattern, Type: null, IsInvalid) (Syntax: 'int z') (InputType: System.Int32?, DeclaredSymbol: System.Int32 z, MatchesNull: False)
+        Value: 
+          ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 5) (Syntax: '5')
+        Locals: Local_1: System.Int32 z
+          Local_2: System.Int32 z
+";
+            var expectedDiagnostics = new[] {
+                // file.cs(7,25): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //         y = /*<bind>*/x switch { int z when x is int z => 5 }/*</bind>*/;
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(7, 25),
+                // file.cs(7,54): error CS0128: A local variable or function named 'z' is already defined in this scope
+                //         y = /*<bind>*/x switch { int z when x is int z => 5 }/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_LocalDuplicate, "z").WithArguments("z").WithLocation(7, 54)
+            };
+            VerifyOperationTreeAndDiagnosticsForTest<SwitchExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/32043")]
```

Comment:
>[Fact(Skip = "https://github.com/dotnet/roslyn/issues/32043")] [](start = 8, length = 62)

Does this test fail locally? Could you describe the failure? It is possible that we can quickly figure out what is wrong.

Reply:
The failure message is "Message: Capture [2] is not used in region [R1] before leaving it after block [B6]" which is an assertion message in the flow graph verifier.
I tried adding `LeaveRegionsUpTo(previousRegion);` (after adjusting the order of operations as you suggested) at the end, but that causes an assertion failure inside that method.

---
In reply to: [244428777](https://github.com/dotnet/roslyn/pull/32018#discussion_r244428777) [](ancestors = 244428777)

Reply:
This is recorded for follow-up at [https://github.com/dotnet/roslyn/issues/32043](https://github.com/dotnet/roslyn/issues/32043)

---
In reply to: [244433542](https://github.com/dotnet/roslyn/pull/32018#discussion_r244433542) [](ancestors = 244433542,244428777)

Reply:
> The failure message is "Message: Capture [2] is not used in region [R1] before leaving it after block [B6]" which is an assertion message in the flow graph verifier.

This is likely about the ```capturedInput```. That capture should be treated as long lived, similar to the expression of a switch statement. A special case should be added to ````bool isLongLivedCaptureReferenceSyntax(SyntaxNode captureReferenceSyntax)``` function in the verifier. Something very similar to the handling for switch:
```
                                case CSharp.SyntaxKind.SwitchStatement:
                                    if (((CSharp.Syntax.SwitchStatementSyntax)syntax.Parent).Expression == syntax)
                                    {
                                        return true;
                                    }
                                    break;

``` 



---
In reply to: [244433725](https://github.com/dotnet/roslyn/pull/32018#discussion_r244433725) [](ancestors = 244433725,244433542,244428777)